pykrige/uk.py

```diff
@@ -820,7 +820,13 @@ def _get_kriging_matrix(self, n, n_withdrifts):
         else:
             a = np.zeros((n_withdrifts, n_withdrifts))
         a[:n, :n] = -self.variogram_function(self.variogram_model_parameters, d)
-        np.fill_diagonal(a, 0.0)
+
+        if self.variogram_model != "linear" and not exact_values:
+            np.fill_diagonal(a, self.variogram_model_parameters[2])

```

Comment:
Could you check, what happens, when you fill the diagonal with the negative nugget? This is what was done in the paper of "3-D interpolation of subsurface temperature data with measurement error using kriging" (Wolfram Rühaak, 2014)
And this is what bothers me: Some literature says you should add the error-variance, other subtract it...

Reply:
Ah, I see your point. I was reading a similar formulation in "Basic Steps in Geostatistics: The Variogram and Kriging" (Oliver and Webster) in Equation 4.11.

![image](https://user-images.githubusercontent.com/10455520/82937057-5f149b80-9f44-11ea-9e01-0e580195b9f4.png)

and comparing this with (Wolfram Rühaak, 2014)

![image](https://user-images.githubusercontent.com/10455520/82937401-f7128500-9f44-11ea-9b98-7425bc1263ba.png)

PyKrige look like its been implemented in the (Wolfram Rühaak, 2014) way, so I think we want to subtract it like you asked, if I'm understanding this correctly. But I _think_ these negatives are redundant.

In the implementation of solving the system to get the linear estimators in `PyKrige` it looks like `b` on the RHS is set to be negative, https://github.com/GeoStat-Framework/PyKrige/blob/1a889a7f087ac7c3e852f6e48601f4a60efca260/pykrige/ok.py#L649 

When finding λ = inv(A) b, the contents of A and b are defined in (Wolfram Rühaak, 2014) and in PyKrige
https://github.com/GeoStat-Framework/PyKrige/blob/1a889a7f087ac7c3e852f6e48601f4a60efca260/pykrige/ok.py#L635, as being negative, meaning λ is positive. 

But in 4.11 in Oliver and Webster, it looks like they've defined the contents of A and b to be positive, which to me seems like the variance/nugget parameter would be added to the diagonal to make the solution for the system of equations (λ = inv(A) b) consistent between both methods, and ensure λ is positive.

So I would expect equivalent behaviour between subtracting the nugget with A and b having negative entries and adding the nugget with A and b having positive entries.

Here is the behaviour from subtracting the nugget, instead, with PyKrige as is:
![PyKrige_NE_Interp](https://user-images.githubusercontent.com/10455520/82948408-94c28000-9f56-11ea-8350-55cbe882bdd8.png)

Which produces some strange variance, perhaps originating from this added term to the variance from Oliver and Webster, Equation 4.12:

![image](https://user-images.githubusercontent.com/10455520/82948884-8032b780-9f57-11ea-9f23-dda5c52205ae.png)

So perhaps https://github.com/GeoStat-Framework/PyKrige/blob/1a889a7f087ac7c3e852f6e48601f4a60efca260/pykrige/ok.py#L622

Needs to have another term if using PyKrige as a non-exact interpolator? In either case, yes, its rather confusing. Does this make sense to you?

Reply:
In GSTools, I was also trying to implement this `exact` feature. There I was following:
Cressie N.: *Statistics for spatial data*
*3.2.1 Effect of Variogram Parameters on Kriging*
Where the measurement error (nugget) was not applied to the kriging matrix, but to the right hand side of the kriging equation. This are the resulting differences:
(exact=True)
![interpolate_exact](https://user-images.githubusercontent.com/19690642/83004838-6d2fee00-a010-11ea-9eaf-041c16d0871d.png)
(exact=False)
![interpolate_nugget](https://user-images.githubusercontent.com/19690642/83004842-6dc88480-a010-11ea-9035-323689ac85eb.png)

And this is what I actually expect from the "exact" keyword, to drop these spikes to the exact conditions and I couldn't reproduce this with one of the above mentioned approaches for the kriging matrix.

I still have to take a look at the estimated kriging variance...

Regarding the formulation: You are right, that the negative sign infront of the variogram doesn't change the outcome. The only advantage of this approach is, that it can be easily reformulated into the system only depending on the covariance instead of the variogram (This is also a goal of mine for PyKrige+GSTools See: https://github.com/GeoStat-Framework/GSTools/issues/79).

And now I know again, why I was struggling with this topic for so long.

Reply:
That behaviour looks like what I would expect, too. My example plots didn't incorporate noise, sorry, so here they are again demonstrating that I _think_ something is missing from that matrix approach:

![PyKrige_pos_nugget](https://user-images.githubusercontent.com/10455520/83056116-b03a9300-a009-11ea-88c3-921ce334ba45.png)

This is for adding a positive nugget to PyKrige's existing sign conventions. It _does not_ produce a reasonable interpolation if the nugget is subtracted.

I would advocate for the RHS approach based on this expected behaviour alone... But its also contradictory to some literature.

Perhaps we might find some insight from existing true and tested implementations of Kriging. I'll see what I can dig up.

Reply:
I would also vote for the RHS approach, since I think, that the argumentation of *Cressie* is the most convincing one we have discussed up to now. I will finish my implementations in GSTools, then we can discuss that, if you are ok with it?

The other approaches always also bring up arguments for the modification of the kriging matrix to gain numerical stability. This issue is also addressed in #151 with the use of the pseudo-inverse matrix.

Reply:
Cressie also gives a hint, that modifying the LHS of the equation is incorrect:
![kriging_note](https://user-images.githubusercontent.com/19690642/83120393-48e81600-a0d1-11ea-9536-e5a43f91c27b.png)

Reply:
And I noticed, that this is in line with H. Wackernagel: *"Multivariate Geostatistics"* (chapter  "Kriging with known measurement error variance"), since there the covariance is used and the jump at the origin of the covariance function is ignored on the LHS.

Reply:
This approach sounds good. In the meantime, I will try and pull up some examples from R's `fields` package, which implementings `mKrig`. I think this could help us verify some behaviour.